Dozuki/CloudPrem-Infra
# Intelligent S3 bucket migration support

# Problem
Our migration system had a fundamental flaw. We could take an existing RDS snapshot and existing S3 buckets to launch a new instance, however that old instance still has a terraform state entry for those S3 buckets. What happens when that old unneeded instance is torn down? The answer is the customer data goes away. 

# Solution
The easiest solution is to simply perform state surgery on that old instance and remove the references to those buckets before deleting and while that works it's far from ideal. One forgotten move or misstep and all of our customer's data is gone forever. I believe this potential calamity is reason enough to invest in a safer replication strategy.

We now create new buckets for each environment regardless of migration state. If migrating, we will setup replication rules on the old buckets and fire off batch replication jobs to copy existing objects from the source buckets to the destination buckets leaving a clear safety separation between environments.

## Caveats
As you may have noticed this has some caveats:
1. S3 bucket replication configurations only apply to **new** objects created after the configuration is applied. To copy **existing** objects, an S3 batch job is required. 
2. We fire off an S3 batch job to copy existing objects but it's asynchronous. It's possible to make this a synchronous process by adding additional logic to the batch creation script but I opted for an async approach due to the unknown and potentially expansive load times some of our larger customer's migrations would take causing TF timeouts and all kinds of messy results. Safety over convenience. So a manual verification of batch job completion is required before the old stack is torn down.
3. Potentially the most annoying caveat and the one the irks me the most is the fact that we have now tied our new state to our old stack (a reversal of what we had before but with zero danger) by creating the replication configuration on the source buckets. That old stack can no longer be torn down fully while the replication configuration is in place (can't delete the bucket when there's still a replication config attached). So to remedy this you would need to re-run this new stack (migration target) with migration disabled so it deletes the replication config thereby freeing up that old stack for deletion. 

Unfortunately there's a bit of a one-two approach to this but I believe on the whole it's a net positive gain for data safety and reliability. 

## Additional Changes
1. To support configuring replication in this way we needed to upgrade to the latest version of the AWS provider. This came with it many refactored resources. The functionality of said resources stays almost the same but the syntax and module implementation slightly differs. 
2. Since we're upgrading the AWS provider we might as well update the other providers we use as well. 
3. Some minor refactors for cleanliness and semantics. Like moving all the data resources out of the `main.tf` into their respective files.
4. Fix replicated install script from trying to re-install on ephemeral environments. 

## QA
I have tested the following scenarios:
* Deploying min, bi, hooks via workstation and cloudformation and validating the app works
* Deploying v3.1 min & full via workstation and cloudformation then upgrading to this branch and verifying nothing bad happens in the TF apply and the app still works
* Deploying this branch, migrating to a new copy of this branch and validating s3 buckets are replicated


Deploying to <= v3.1 and migrating to a new copy of this branch is not supported. Update the source instance first before migration.

Some potential issues that cropped up:
* if we upgrade a customer from <=v3.1 to this version then the ~database passwords~, self-signed certificates, and s3 KMS key will be changed. 
  * ~This requires a full reboot of the application with some minor downtime (<5 minutes) to load the new db credentials.~ (Fixed in 2fb7b5de8fbb07dae000ac13d025f047cab79dc2)
  * ~The bastion instance will need to be recycled to load the new db credentials.~
  * If this new upgraded stack were to ever be migrated in the future someone would need to run a batch job to normalize all the objects in the buckets to use the same key since any objects created before this update will be using the default KMS key and any after will be using the bespoke key.
